docs/strings.md

```diff
@@ -0,0 +1,464 @@
+
+There are two other uncommon Unicode encodings that are worth mentioning but probably not worth supporting as an internal representation for standard library strings:
+
+ - UTF-7 is an encoding into a 7-bit character space for use with old email systems that cannot process 8-bit text.
+ - UTF-EBCDIC is equivalent to UTF-8 but stores the EBCDIC character set in the 7-bit space, rather than the ASCII set.
+
+There is no single encoding that is the best fit for all applications.
+There are two key design questions related to encodings:
+
+ - Should strings that use a non-Unicode internal encoding ever expose that across the API boundary?
+   For example, should Japanese text in Big5 encoding be required to convert to Unicode for all operations that are not part of the class's internal implementation?
+ - Should strings that are using a Unicode encoding expose interfaces in a *specific* unicode encoding?
+
+*Proposal*: Standard library strings should expose interfaces in 32-bit Unicode code points.
+They should also be able to export or provide views in their native encoding so that operations can be optimised based on pattern matching.
+
+Draft proposal for a core string library
+----------------------------------------
+
+The Verona standard string should define a string interface and a small set of concrete instantiations of this interface.
+The core string class should implement the minimum required for generic implementation of string operations, which can be implemented either using dynamic dispatch over the string interface or as generic reifications over the concrete types.
+We should make it possible to support a rich Unicode library that uses the same string objects as the standard library (and third-party libraries that use the same interface) and provides functions for iteration with unicode graphemes, locale-aware collation and other more complex operations that not all programs need.
+
+The interface for strings would look roughly like this:
+
+```verona
+/**
+ * String contains only 7-bit ASCII characters and is stored with each code
+ * point in an 8-bit integer.
+ */
+class StringEncodingASCII
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-8-encoded text.
+ */
+class StringEncodingUTF8
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-16-encoded text.
+ */
+class StringEncodingUTF16
+{
+	typedef CharType = U16;
+}
+/**
+ * String is stored as UTF-32-encoded text.
+ */
+class StringEncodingUTF32
+{
+	typedef CharType = U32;
+}
+/**
+ * Interface for other string encodings.  Any other type can be used as a valid
+ * string encoding but code that does not have any special cases for that
+ * encoding will be able to treat is as unknown.
+ */
+interface StringEncodingCustom
```

Comment:
Why `Custom` and not just `interface StringEncoding`?  The `class`es above ascribe to this `interface`, yes?

Reply:
I think we want to differentiate between text stored internally as unicode and something that is custom.  My rationale for this was to make it easy to pattern match on 'shrug, go to generic slow paths' vs 'unicode internally, we can do something fast here'.  This may not actually be useful.